### PR TITLE
refactor model, forwardproblem, backwardproblem

### PR DESCRIPTION
**Pull request by FFroehlich**

### CODE REVIEW

File: include/amici/misc.h [hunk 1]
```diff
@@ -11,6 +11,17 @@
...
+    
+/**
+ * @brief creates a slice from existing data
+ *
+ * @param data to be sliced
+ * @param index slice index
+ * @param size slice size
+ */
+ 
+ gsl::span<realtype> slice(std::vector<realtype> data, const int index,
```
Comment (dweindl) [inline include/amici/misc.h]: See also gsl-lite.hpp: `span subspan( index_type offset, index_type count )`

File: src/model.cpp [hunk 2]
```diff
@@ -14,6 +14,26 @@
...
 namespace amici {
 
+static void checkBufferSize(gsl::span<realtype> buffer,
+                            __SIZE_TYPE__ expected_size) {
+    if (buffer.size() != static_cast<size_t>(expected_size))
+        throw AmiException("Incorrect buffer size! Was %i, expected %i.",
+                           buffer.size(), expected_size);
+}
+
+static void writeSlice(std::vector<realtype> slice,
```
Comment (dweindl) [inline src/model.cpp]:
```suggestion
static void writeSlice(gsl::span<const realtype> slice,
```

File: src/model.cpp [hunk 3]
```diff
@@ -14,6 +14,26 @@
...
+                       gsl::span<realtype> buffer) {
+    checkBufferSize(buffer, slice.size());
+    std::copy(slice.begin(), slice.end(), buffer.data());
+}
+
+static void reset_storage(std::vector<realtype> vec, const int size) {
```
Comment (dweindl) [inline src/model.cpp]: `void std::assign (size_type n, const value_type& val);`

Should do exactly what those 4 lines do.

File: src/model.cpp [hunk 4]
```diff
@@ -71,6 +78,17 @@ Model::Model(const int nx_rdata, const int nxtrue_rdata, const int nx_solver,
     requireSensitivitiesForAllParameters();
 }
 
+void Model::checkLLHBufferSize(std::vector<realtype> &sllh,
+                               std::vector<realtype> &s2llh) {
+    if (sllh.size() != nplist())
+        throw AmiException("Incorrect sllh buffer size! Was %i, expected %i.",
```
Comment (dweindl) [inline src/model.cpp]: Won't print the correct sizes, I think. `%u`

File: include/amici/misc.h (hunk 1 above)
Comment (FFroehlich) [reply include/amici/misc.h]: subspan does not multiply the offset with size.

File: include/amici/model.h [hunk 5]
```diff
@@ -1406,126 +1518,206 @@ class Model : public AbstractModel {
...
+
+    /** parameter derivative of data likelihood for current timepoint
+     * (dimension: nJ x nplist, row-major)
+     */
+    mutable std::vector<realtype> dJydp;
 
-    /** observable derivative of data likelihood (dimension nJ x nytrue x ny,
-      * ordering = ?) (only used if wasPythonGenerated()==false ) */
-    std::vector<realtype> dJydy_matlab;
+    /** event ouput derivative of event likelihood
```
Comment (dweindl) [inline include/amici/model.h]:
```suggestion
    /** event output derivative of event likelihood
```

File: include/amici/model.h [hunk 6]
```diff
@@ -1406,126 +1518,206 @@ class Model : public AbstractModel {
...
-    std::vector<realtype> dJydsigma;
+    /** event sigma derivative of event likelihood
+     * (dimension nJ x nz x nztrue, row-major)
+     */
+    mutable std::vector<realtype> dJzdsigma;
 
-    /** event ouput derivative of event likelihood (dimension nJ x nztrue x nz,
-     * ordering = ?) */
-    std::vector<realtype> dJzdz;
+    /** event ouput derivative of event likelihood at final timepoint
```
Comment (dweindl) [inline include/amici/model.h]:
```suggestion
    /** event output derivative of event likelihood at final timepoint
```

File: include/amici/model.h [hunk 7]
```diff
@@ -1406,126 +1518,206 @@ class Model : public AbstractModel {
...
 
-    /** state derivative of observable
-     * (dimension: nx_solver x ny, ordering = row-major) */
-    std::vector<realtype> dydx;
+    /** state derivative of time-resolved observable
+     * (dimension: nx_solver x ny, row-major)
+     */
+    mutable std::vector<realtype> dydx;
 
     /** tempory storage of w data across functions (dimension: nw) */
```
Comment (dweindl) [inline include/amici/model.h]: s/tempory/temporary/

**Pull request merged**